src/Features/CSharp/Portable/ExtractMethod/CSharpMethodExtractor.cs

```diff
@@ -24,24 +24,30 @@ namespace Microsoft.CodeAnalysis.CSharp.ExtractMethod
     internal partial class CSharpMethodExtractor(CSharpSelectionResult result, ExtractMethodGenerationOptions options, bool localFunction)
         : MethodExtractor(result, options, localFunction)
     {
-        protected override Task<AnalyzerResult> AnalyzeAsync(SelectionResult selectionResult, bool localFunction, CancellationToken cancellationToken)
-            => CSharpAnalyzer.AnalyzeAsync(selectionResult, localFunction, cancellationToken);
+        protected override AnalyzerResult Analyze(SelectionResult selectionResult, bool localFunction, CancellationToken cancellationToken)
+            => CSharpAnalyzer.Analyze(selectionResult, localFunction, cancellationToken);
 
-        protected override async Task<InsertionPoint> GetInsertionPointAsync(SemanticDocument document, CancellationToken cancellationToken)
+        protected override SyntaxNode GetInsertionPointNode(
+            AnalyzerResult analyzerResult, CancellationToken cancellationToken)
         {
             var originalSpanStart = OriginalSelectionResult.OriginalSpan.Start;
             Contract.ThrowIfFalse(originalSpanStart >= 0);
 
-            var root = await document.Document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
-            var basePosition = root.FindToken(originalSpanStart);
+            var document = this.OriginalSelectionResult.SemanticDocument;
+            var root = document.Root;
 
             if (LocalFunction)
             {
                 // If we are extracting a local function and are within a local function, then we want the new function to be created within the
                 // existing local function instead of the overarching method.
+                var outermostCapturedVariable = analyzerResult.GetOutermostVariableToMoveIntoMethodDefinition(cancellationToken);
+                var baseNode = outermostCapturedVariable != null
+                    ? outermostCapturedVariable.GetIdentifierTokenAtDeclaration(document).Parent
+                    : this.OriginalSelectionResult.GetOutermostCallSiteContainerToProcess(cancellationToken);
```

Comment:
there was a bunch of problems with "Extract local function" in that we computed varying different containers at different points in the algorithm, leading to the actual rewrite later on sometimes replacing hte node that extract local function wanted to insert the local function into.  We compensated for this by adding a later check and bailing out.  But that won't work if we want to defer generation to code action commit.

So, instead, i've pulled back the syntactic work to actually determine what is going to be rewritten, and now when we're figuring out where the new local function should go, we take that into account here.

Reply:
The logic for figuring out hte container for the 'exract local function' vs 'extract method' cases are extremely different.  i ended up just separating them out completely to make things clearer.